scenarios/smart_trailer/digital_twin_providers/trailer_properties_provider/src/main.rs

```diff
@@ -0,0 +1,170 @@
+const PROVIDER_AUTHORITY: &str = "0.0.0.0:4030";
+
+const DEFAULT_MIN_INTERVAL_MS: u64 = 10000; // 10 seconds
+
+/// Register the trailer weight property's endpoint.
+///
+/// # Arguments
+/// * `invehicle_digital_twin_uri` - The In-Vehicle Digital Twin URI.
+/// * `provider_uri` - The provider's URI.
+async fn register_trailer_weight(
+    invehicle_digital_twin_uri: &str,
+    provider_uri: &str,
+) -> Result<(), Status> {
+    let endpoint_info = EndpointInfo {
+        protocol: digital_twin_protocol::GRPC.to_string(),
+        operations: vec![digital_twin_operation::MANAGEDSUBSCRIBE.to_string()],
+        uri: provider_uri.to_string(),
+        context: "GetSubscriptionInfo".to_string(),
+    };
+
+    let entity_access_info = EntityAccessInfo {
+        name: trailer_v1::trailer::trailer_weight::NAME.to_string(),
+        id: trailer_v1::trailer::trailer_weight::ID.to_string(),
+        description: trailer_v1::trailer::trailer_weight::DESCRIPTION.to_string(),
+        endpoint_info_list: vec![endpoint_info],
+    };
+
+    let mut client = InvehicleDigitalTwinClient::connect(invehicle_digital_twin_uri.to_string())
+        .await
+        .map_err(|e| Status::internal(e.to_string()))?;
+    let request = tonic::Request::new(RegisterRequest {
+        entity_access_info_list: vec![entity_access_info],
+    });
+    let _response = client.register(request).await?;
+
+    Ok(())
+}
+
+/// Start the trailer weight data stream.
+///
+/// # Arguments
+/// `min_interval_ms` - minimum frequency for data stream.
+fn start_trailer_weight_data_stream(min_interval_ms: u64) -> watch::Receiver<i32> {
+    debug!("Starting the Provider's trailer weight data stream.");
+    let mut weight: i32 = 1000;
+    let (sender, reciever) = watch::channel(weight);
+    tokio::spawn(async move {
+        let mut is_weight_increasing: bool = true;
+        loop {
+            debug!(
+                "Recording new value for {} of {weight}",
+                trailer_v1::trailer::trailer_weight::ID
+            );
+
+            if let Err(err) = sender.send(weight) {
+                warn!("Failed to get new value due to '{err:?}'");
+                break;
+            }
+
+            debug!("Completed the publish request");
+
+            // Calculate the new weight.
+            // It bounces back and forth between 1000 and 2000 kilograms.
+            // It increases in increments of 500 to simulate a large amount of cargo being loaded
+            // And decreases in increments of 50 to simulate smaller deliveries being made
+            if is_weight_increasing {
+                if weight == 2000 {
+                    is_weight_increasing = false;
+                    weight -= 50;
+                } else {
+                    weight += 500;
+                }
+            } else if weight == 1000 {
+                is_weight_increasing = true;
+                weight += 500;
+            } else {
+                weight -= 50;
+            }
```

Comment:
The current implementation works fine, but it might be a bit easier to read with less nesting/branching with one of the following:

```rust
if is_weight_increasing {
    weight += 50;
} else {
    weight -= 50;
}

if weight == 2000 {
    is_weight_increasing = false;
} else if weight == 1000 {
    is_weight_increasing = true;
}
```

```rust
if is_weight_increasing && weight < 2000 || weight == 1000 {
    is_weight_increasing = true;
    weight += 50;
} else {
    is_weight_increasing = false;
    weight -= 50;
}
```

Reply:
may also be worth converting the weight limits to constants (e.g. `MIN_WEIGHT` and `MAX_WEIGHT`)

Reply:
either way, looks like there's a typo on line 102, should be 50 instead of 500

Reply:
I like the third option you showed